# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: caab023e5d2d
# Full Hash: caab023e5d2d6daee7993353cecc142dde2483fb
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2024-03-11 21:13:39
# Regressor Bug: 1860328
# File Overlap Count: 2
# Description:
#   Bug 1860328 - Fix a bug where caret misses when dragging and dropping r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D192377
# ==============================================================================

diff -r 09b0c2bfb164 -r caab023e5d2d layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Mon Mar 11 13:36:59 2024 +0000
+++ b/layout/painting/nsDisplayList.cpp	Mon Mar 11 13:41:27 2024 +0000
@@ -650,7 +650,6 @@
       mCurrentContainerASR(nullptr),
       mCurrentFrame(aReferenceFrame),
       mCurrentReferenceFrame(aReferenceFrame),
-      mCaretFrame(nullptr),
       mScrollInfoItemsForHoisting(nullptr),
       mFirstClipChainToDestroy(nullptr),
       mTableBackgroundSet(nullptr),
@@ -718,21 +717,6 @@
       mRetainingDisplayList && StaticPrefs::layout_display_list_retain_sc();
 }
 
-static PresShell* GetFocusedPresShell() {
-  nsPIDOMWindowOuter* focusedWnd =
-      nsFocusManager::GetFocusManager()->GetFocusedWindow();
-  if (!focusedWnd) {
-    return nullptr;
-  }
-
-  nsCOMPtr<nsIDocShell> focusedDocShell = focusedWnd->GetDocShell();
-  if (!focusedDocShell) {
-    return nullptr;
-  }
-
-  return focusedDocShell->GetPresShell();
-}
-
 void nsDisplayListBuilder::BeginFrame() {
   nsCSSRendering::BeginFrameTreesLocked();
 
@@ -742,26 +726,6 @@
   mInTransform = false;
   mInFilter = false;
   mSyncDecodeImages = false;
-
-  if (!mBuildCaret) {
-    return;
-  }
-
-  RefPtr<PresShell> presShell = GetFocusedPresShell();
-  if (presShell) {
-    RefPtr<nsCaret> caret = presShell->GetCaret();
-    mCaretFrame = caret->GetPaintGeometry(&mCaretRect);
-
-    // The focused pres shell may not be in the document that we're
-    // painting, or be in a popup. Check if the display root for
-    // the caret matches the display root that we're painting, and
-    // only use it if it matches.
-    if (mCaretFrame &&
-        nsLayoutUtils::GetDisplayRootFrame(mCaretFrame) !=
-            nsLayoutUtils::GetDisplayRootFrame(mReferenceFrame)) {
-      mCaretFrame = nullptr;
-    }
-  }
 }
 
 void nsDisplayListBuilder::AddEffectUpdate(dom::RemoteBrowser* aBrowser,
@@ -801,7 +765,6 @@
   FreeClipChains();
   FreeTemporaryItems();
   nsCSSRendering::EndFrameTreesLocked();
-  mCaretFrame = nullptr;
 }
 
 void nsDisplayListBuilder::MarkFrameForDisplay(nsIFrame* aFrame,
@@ -1141,11 +1104,27 @@
     return;
   }
 
+  RefPtr<nsCaret> caret = state->mPresShell->GetCaret();
+  // This code run for each pres shell and caret->GetPaintGeometry
+  // will return nullptr for invisible caret. So only one caret
+  // can be painted at a time.
+  state->mCaretFrame = caret->GetPaintGeometry(&mCaretRect);
+
+  // Check if the display root for the caret matches the display
+  // root that we're painting, and only use it if it matches. Likely
+  // we only need this for popup.
+  if (state->mCaretFrame &&
+      nsLayoutUtils::GetDisplayRootFrame(state->mCaretFrame) !=
+          nsLayoutUtils::GetDisplayRootFrame(aReferenceFrame)) {
+    state->mCaretFrame = nullptr;
+  }
+
   // Caret frames add visual area to their frame, but we don't update the
   // overflow area. Use flags to make sure we build display items for that frame
   // instead.
-  if (mCaretFrame && mCaretFrame->PresShell() == state->mPresShell) {
-    MarkFrameForDisplay(mCaretFrame, aReferenceFrame);
+  if (state->mCaretFrame) {
+    MOZ_ASSERT(state->mCaretFrame->PresShell() == state->mPresShell);
+    MarkFrameForDisplay(state->mCaretFrame, aReferenceFrame);
   }
 }
 
@@ -4134,8 +4113,8 @@
   nscolor caretColor;
   nsIFrame* frame =
       mCaret->GetPaintGeometry(&caretRect, &hookRect, &caretColor);
-  MOZ_ASSERT(frame == mFrame, "We're referring different frame");
-  if (!frame) {
+  if (NS_WARN_IF(!frame) || NS_WARN_IF(frame != mFrame)) {
+    NS_ASSERTION(false, "Caret invalidation bug");
     return true;
   }
 